### PR TITLE
#42 - Handle lists enclosed by some characters

### PR DESCRIPTION
**Pull request by calston**
This adds an `ends_with` optional parameter to accept a character or sub-string which overrides the existing logic and instead simply indexes the end of the list (based on the ends_with match in revers [rindex]) rather than the string end or delimiter end allowing the modification of lists which are quoted, or bracketed.

### CODE REVIEW

**Review by steflsd — COMMENTED** [review]
Tested and functionally does what I need. I would suggest this should be part of the core release asap.

File: libraries/provider_add_to_list.rb [hunk 1]
```diff
@@ -48,17 +48,25 @@ def action_edit
             f.each_line do |line|
               if line =~ regex
                 found = true
-                if new_resource.delim.count == 1
-                  unless line =~ /(#{new_resource.delim[0]}|#{new_resource.pattern})\s*#{new_resource.entry}\s*(#{new_resource.delim[0]}|\n)/
-                    line = line.chomp + "#{new_resource.delim[0]}#{new_resource.entry}"
+                if new_resource.ends_with
+                  unless line =~ /(#{new_resource.delim[0]}|#{new_resource.pattern}).+#{new_resource.entry}(#{new_resource.delim[0]}|#{new_resource.ends_with}|\n)/
+                    list_end = line.rindex(new_resource.ends_with)
+                    line = line.chomp.insert(list_end, "#{new_resource.delim[0]}#{new_resource.entry}")
```
Comment (MarkGibbons) [inline libraries/provider_add_to_list.rb:54]: list_end is nil if the line doesn't match new_resource.ends_with. I think the regex should check for ends_with being present.  In cases where ends_with includes ) or ] the values also need to be escaped.

File: libraries/provider_add_to_list.rb [hunk 2]
```diff
@@ -48,17 +48,25 @@ def action_edit
             f.each_line do |line|
               if line =~ regex
                 found = true
-                if new_resource.delim.count == 1
-                  unless line =~ /(#{new_resource.delim[0]}|#{new_resource.pattern})\s*#{new_resource.entry}\s*(#{new_resource.delim[0]}|\n)/
-                    line = line.chomp + "#{new_resource.delim[0]}#{new_resource.entry}"
+                if new_resource.ends_with
+                  unless line =~ /(#{new_resource.delim[0]}|#{new_resource.pattern}).+#{new_resource.entry}(#{new_resource.delim[0]}|#{new_resource.ends_with}|\n)/
```
Comment (MarkGibbons) [inline libraries/provider_add_to_list.rb:52]: The check is for #{new_resource.entry}(#{new_resource.delim[0]}  shouldn't the check be for delim[0] before the new entry.

Comment (calston) [reply libraries/provider_add_to_list.rb:52]: The regex should render out to `..."foo(,|;|\n)"` (with , as the delimiter and ; as the end stop). Which should match `foo;` and `foo, bar` and ... uh... `foo\n` which isn't really valid but what do you do in that case anyway?

That said I definitely think there's a much more elegant and readable algorithm to parse stuff like this without having complicated regexes. When I try to follow the existing ones I think there are tons of edge cases they will silently fail on and not raise an exception in the recipe, and the corresponding tests definitely need improvement.

**Comment by someara** [comment]
thanks!
